File: windows/src/SkillHub.Core/HubError.cs
namespace SkillHub.Core;

public enum HubErrorKind
{
    AccessDenied,
    PathNotAllowed,
    BuiltinReadOnly,
    AlreadyExists,
    NotFound,
    InvalidName,
    LastEntity,
    Io
}

public readonly record struct HubError(HubErrorKind Kind, string Message, string? Detail = null)
{
    public static HubError AccessDenied() =>
        new(HubErrorKind.AccessDenied, "没有拿到主目录的访问权限。");

    public static HubError PathNotAllowed(string path) =>
        new(HubErrorKind.PathNotAllowed, $"路径不在允许的根目录内：{path}", path);

    public static HubError BuiltinReadOnly() =>
        new(HubErrorKind.BuiltinReadOnly, "Cursor 内置 skill 只读，不能修改。");

    public static HubError AlreadyExists(string name) =>
        new(HubErrorKind.AlreadyExists, $"「{name}」已经存在。", name);

    public static HubError NotFound() =>
        new(HubErrorKind.NotFound, "找不到这个文件。");

    public static HubError InvalidName() =>
        new(HubErrorKind.InvalidName, "名字只能用小写字母、数字和连字符，最多 64 个字符。");

    public static HubError LastEntity() =>
        new(HubErrorKind.LastEntity, "这是最后一份实体文件，不能只当作链接移除。");

    public static HubError Io(string message) =>
        new(HubErrorKind.Io, message);
}

public sealed class HubException : Exception
{
    public HubError Error { get; }

    public HubException(HubError error) : base(error.Message)
    {
        Error = error;
    }
}

File: SkillHub/Catalog/Models/HubError.swift
import Foundation

nonisolated enum HubError: LocalizedError, Equatable {
    case accessDenied
    case pathNotAllowed(String)
    case builtinReadOnly
    case alreadyExists(String)
    case notFound
    case invalidName
    case lastEntity
    case io(String)

    var errorDescription: String? {
        switch self {
        case .accessDenied:
            String(localized: "没有拿到主目录的访问权限。请在「文件 › 重新授权主目录…」里重新授权。")
        case .pathNotAllowed(let path):
            String(localized: "路径不在允许的根目录内：\(path)")
        case .builtinReadOnly:
            String(localized: "Cursor 内置 skill 只读，不能修改。")
        case .alreadyExists(let name):
            String(localized: "「\(name)」已经存在。")
        case .notFound:
            String(localized: "找不到这个文件。")
        case .invalidName:
            String(localized: "名字只能用小写字母、数字和连字符，最多 64 个字符。")
        case .lastEntity:
            String(localized: "这是最后一份实体文件，不能只当作链接移除。")
        case .io(let message):
            message
        }
    }
}
